File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverse(ListNode head){
        ListNode p1=null, p2=head;
        while(p2!=null){
            ListNode temp = p2.next;
            p2.next = p1;
            p1=p2;
            p2=temp;
        }
        return p1;
    }
    public ListNode removeNthFromEnd(ListNode head, int n) {
        head = reverse(head);
        if(head.next==null && (n==1) )return null;
        ListNode p1=head, p2=head.next;
        if(n==1){
            head = head.next;
        }else{
            while( (n-1) > 1){
                p1 = p1.next;
                p2 = p2.next;
                n--;
            }
            p1.next = p2.next;
        }
        head = reverse(head);
        return head;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        first, second = head, head.next
        while second and second.next:
            first  = first.next
            second = second.next.next
        second = first.next

        prev = first.next = None ####
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp
        # first, second

        first = head
        second = prev
        while second:
            tmp1 = first.next
            tmp2 = second.next
            first.next = second
            second.next = tmp1
            first = tmp1
            second = tmp2

        # dn = p = ListNode()
        # x, y = first, second
        # # while x and y:
        # while y:
        #     # p.next = x
        #     p = x
        #     x = x.next
        #     p.next = y
        #     p = y
        #     y = y.next
        # p.next = x


